src/dependency-manager/manager.ts

```diff
@@ -361,16 +361,91 @@ export default abstract class DependencyManager {
+        ...this.validateComponentResourceSpecs(component_spec.services, secrets, component_spec.name, 'services'),
+        ...this.validateComponentResourceSpecs(component_spec.tasks, secrets, component_spec.name, 'tasks'),
+      ];
+      if (validation_errors.length > 0) {
+        throw new ValidationErrors(validation_errors, component_spec.metadata.file);
+      }
+    }
+  }
+
+  updateResourceEnvironmentSecrets(resource_specs: Dictionary<ResourceSpec> = {}, all_secrets: SecretsDict, component_name: string): Dictionary<ResourceSpec> {
+    for (const [resource_name, resource_spec] of Object.entries(resource_specs)) {
+      const resource_environment = resource_specs[resource_name].environment;
+
+      if (resource_environment) {
+        for (const [env_var_key, env_var_value] of Object.entries(resource_spec.environment || {})) {
+          if (typeof resource_environment[env_var_key] !== 'string') { // don't overwrite hardcoded environment variables or interpolation references
+            if (all_secrets[component_name] && all_secrets[component_name][env_var_key]) {
+              resource_environment[env_var_key] = all_secrets[component_name][env_var_key];
+            } else if (all_secrets['*'] && all_secrets['*'][env_var_key]) {
+              resource_environment[env_var_key] = all_secrets['*'][env_var_key];
+            } else if (env_var_value && typeof env_var_value === 'object') {
+              const secret_definition_spec = env_var_value as SecretDefinitionSpec;
+              if (secret_definition_spec.default) {
+                resource_environment[env_var_key] = secret_definition_spec.default || null;
+              } else if (secret_definition_spec.required === false) {
+                resource_environment[env_var_key] = null; // no matching secret passed in, environment variable optional
+              }
+            }
+          }
+        }
+      }
+    }
+
+    return resource_specs;
+  }
+
+  updateResourceEnvironments(component_specs: ComponentSpec[], all_secrets: SecretsDict): ComponentSpec[] {
+    const updated_component_specs = [];
+    for (const component_spec of component_specs) {
+      const updated_component_spec = component_spec;
+      updated_component_spec.services = this.updateResourceEnvironmentSecrets(component_spec.services, all_secrets, component_spec.name);
+      updated_component_spec.tasks = this.updateResourceEnvironmentSecrets(component_spec.tasks, all_secrets, component_spec.name);
+      updated_component_specs.push(updated_component_spec);
+    }
+    return updated_component_specs;
+  }
+
+  async getGraph(original_component_specs: ComponentSpec[], all_secrets: SecretsDict = {}, options?: GraphOptions): Promise<DependencyGraph> {
+    options = {
       interpolate: true,
       validate: true,
       ...options,
     };
 
     if (options.validate) {
       SecretsConfig.validate(all_secrets);
+      this.validateEnvironments(original_component_specs, all_secrets);
```

Comment:
Making a note to myself that it'd prob be easier to validate the ComponentConfig than the ComponentSpec. Will come back to this.